paramiko/pkey.py

```diff
@@ -39,6 +39,132 @@
 from paramiko.message import Message
 
 
+_legacy_type_registry = {}
+_openssh_type_registry = []
+
+
+def register_pkey_type(cls):
+    """Decorator for PKey subclasses to register their types for parsing."""
+    if cls.LEGACY_TYPE:
+        _legacy_type_registry[cls.LEGACY_TYPE] = cls
+    if cls.OPENSSH_TYPE_PREFIX:
+        _openssh_type_registry.append((cls.OPENSSH_TYPE_PREFIX, cls))
+    return cls
+
+
+# TODO: look at what documentation needs to be written
+# TODO: look for code in paramiko-ng internally that could use this
+# TODO: tests for "wrong key type"
+# TODO: check code style with Black?
+def load_private_key(key_str, password=None):
+    """
+    Autodetect the private key type and load it from a string.
+    Returns some subclass of PKey.
+
+    .. versionadded:: 2.7
+    """
+    pkformat, typ, data = _read_private_key(key_str, password)
+
+    if pkformat == PKey.FORMAT_ORIGINAL:
+        cls = _legacy_type_registry.get(typ)
+    elif pkformat == PKey.FORMAT_OPENSSH:
+        cls = None
+        for map_type, map_cls in _openssh_type_registry:
+            if typ.startswith(map_type):
+                cls = map_cls
+    else:
+        assert False, "Unexpected pkformat %s" % pkformat
+
+    if cls is None:
+        raise SSHException("Unsupported key type %s" % typ)
+
+    return cls(raw=(pkformat, data))
+
+
+def _read_private_key(key_str, password):
+    pkformat, typ, headers, data = _parse_openssh_pkey(key_str)
+    if pkformat == PKey.FORMAT_ORIGINAL:
+        data = _read_private_key_old_format(headers, data, password)
+    elif pkformat == PKey.FORMAT_OPENSSH:
+        assert not headers
+        typ, data = _read_private_key_new_format(data, password)
+    return pkformat, typ, data
+
+
+def load_private_key_file(filename, password=None):
```

Comment:
Mainly two new public APIs were added: `load_private_key` (string API) and `load_private_key_file`.

The `load_private_key_file_obj` was dropped because I don't think it's needed. A user can simply do the `.read()` themselves.

Reply:
👍